### ISSUE TITLE
[Question] When the Application in IIS stops, what happens to the running jobs

### ISSUE THREAD

**Issue by farhadnowzari** [jamesmh/coravel#228]
Hello everyone,

thank you for the project.

I have a question which I know its not the place to ask but looks like an easy shortcut ;)

When a job from a scheduler is running I saw that if in the middle of that running job I stop the IIS application for this website, Coravel will not dispose and instead it will log that it will wait until the schedule is done but it seems it never finishes the job.
Coravel never got to this part with this scenario.

https://github.com/jamesmh/coravel/blob/751d1c39fad0f65b8201d30b64a3b25053818e6d/Src/Coravel/Scheduling/HostedService/SchedulerHost.cs#L60

and I've also never got the result I was waiting for to get from my job.
I know I maybe expecting a magical thing here because obviously if the IIS app is stopped then the scheduler is stopped immediately even though it says it is not (because it is running inside the stopped IIS application), am I right or not?

**Comment by jamesmh** [comment]
If you have a job that takes long enough to halt the application from shutting down (Coravel will do everything it can to prevent the IIS application from fully stopping when there are still active jobs) then [you need to manage the cancellation of these kinds of long-running jobs manually](https://docs.coravel.net/Scheduler/#invocables). Otherwise, Coravel will try to keep running as long as IIS will allow it - even if, let's say, your database is not available, etc.

![image](https://user-images.githubusercontent.com/7213174/120335760-5a9d6400-c2b7-11eb-9b66-2945cfda19e3.png)

By using the cancellation token your long-running jobs can detect when the system is going down and gracefully stop the job and potentially resume next time the scheduled task is enabled.

Let me know how that works out 👍

**Comment by farhadnowzari** [comment]
Hi,
Thanks for your answer.
I'm not quit sure if I was able to describe my question very well but this cancellation is kinda related but I have not thought about it at this point.

But about the question, please consider a job as follow which is scheduled to be ran every lets say six minutes without overlaps.
```
    public class TestJob : IInvocable
    {
        public async Task Invoke()
        {
            var jobIdentifier = Guid.NewGuid();
            File.AppendAllText($"D:\\{jobIdentifier.ToString()}", DateTime.UtcNow.ToString());
            await Task.Delay(60000);
            File.AppendAllText($"D:\\{jobIdentifier.ToString()}", DateTime.UtcNow.ToString());
        }
    }
```
And as you see, It waits between the first file write and the second one. I know you have mentioned that Coravel will make sure that the job runs when for example I stop IIS for a deployment.
But the thing is that when I test it and stop the application like below in IIS while the job is waiting in between these two writes in my sample code, it doesn't continue running the job. I waited for it, So it writes into the file again but it didn't. 
I tested it without stopping IIS and it was fine. 
It can be an IIS config which I'm not familiar with maybe. It seems Coravel is logging the correct message about waiting until the job is done but the job never ends after I stop the IIS app, which should because this job is not complicated after and one minute I have to see the second write.

![image](https://user-images.githubusercontent.com/57224511/120359847-6f561800-c308-11eb-8cc9-bcddf0eb178a.png)

**Comment by jamesmh** [comment]
Yes, IIS has it's own internal timeout that will kill the .NET process altogether if it doesn't stop. So what's probably happening is:

1. IIS is requested to stop
2. .NET tells IIS "I still have work to do...just wait!"
3. IIS waits for .NET to stop
4. IIS waits longer than it's internally configured timeout
5. IIS kills the .NET processes completely

So at step 5, IIS just kills the process. So in effect,  your job is cancelled forcefully by IIS. This is why you need to make sure you handle the cancellation yourself in a graceful manner.

**Comment by farhadnowzari** [comment]
Ok, I will try now. The idle timeout is set to its default, 20mins but I think maybe the Stop LinkButton in IIS force stop the application and kills the process
which means it makes the deployment a bit interesting.
I will try to see what happens.

**Comment by jamesmh** [comment]
Awesome. This is an issue I see quite often. It's part of the territory, so the speak lol. Jobs need to either be fast to finish or if they are longer running need to be able to gracefully stop in the middle of processing.

If that's absolutely not possible, many choose make their scheduled jobs emit a command to a message queue (like rabbitmq) that will trigger the job to run. The command/message will be consumed by another worker service, not tied to the web site, and then is able to do the work. You lose some of coravel's features, like the prevent overlapping in this case. But, if you de-couple the deployment of the web site and the worker then you can deploy the web site while the worker is unaffected. Just one other option in the grand scheme.

**Comment by farhadnowzari** [comment]
hmm, that sounds considerable and of course running a part of the application on a service causes that for example the data layer to be shared between the two and for example these Services/Microservices access the same database.

btw I figured out something with IIS.
The idle timeout works perfectly, it seems Coravel will not let the IIS to terminate untils it is done but if you do the stop yourself on IIS it kills the process and the progress is gone.
So yes I think we need to think this through.

I've had experience with Laravel Horizon which runs the jobs on Linux Cron scheduler and keeps the code on the same app. Then in prod when we wanted to deploy we just needed to stop the docker container which was running the website while the running jobs where still running unless we needed to force stop them.

But great job with Coravel library. I appreciate the effort you've put into it.
